simSPI/linear_simulator/projector.py

```diff
@@ -21,18 +21,144 @@ def __init__(self, config):
+            # torch.nn.functional.grid_sample
+            coords = 2 * coords
+            self.register_buffer("vol_coords", coords.reshape(-1, 2))
+        else:
+            raise NotImplementedError(
+                f"Space type '{self.space}' " f"has not been implemented!"
+            )
 
     def forward(self, rot_params, proj_axis=-1):
+        """Forward method for projection.
+
+        Parameters
+        ----------
+        rot_params : tensor of rotation matrices
+        """
+        if self.space == "real":
+            return self._forward_real(rot_params, proj_axis)
+
+        if self.space == "fourier":
+            if proj_axis != -1:
+                raise NotImplementedError(
+                    "proj_axis must currently be -1 for Fourier space projection"
+                )
+            return self._forward_fourier(rot_params)
+        raise NotImplementedError(
+            f"Space type '{self.space}' " f"has not been implemented!"
+        )
+
+    def _forward_fourier(self, rot_params):
+        """Output the tomographic projection of the volume in Fourier space.
+
+        Take a slide through the Fourier space volume whose normal is
+        oriented according to rot_params.  The volume is assumed to be cube
+        represented in the fourier space.  The output image follows
+        (batch x channel x height x width) convention of pytorch.  Therefore,
+        a dummy channel dimension is added at the end to projection.
+
+        Parameters
+        ----------
+        rot_params: dict of type str to {tensor}
+            Dictionary containing parameters for rotation, with keys
+                rotmat: str map to tensor
+                    rotation matrix (batch_size x 3 x 3) to rotate the volume
+
+        Returns
+        -------
+        projection: tensor
+            Tensor containing tomographic projection in the Fourier domain
+            (batch_size x 1 x sidelen x sidelen)
+
+        Comments
+        --------
+        Note that the Fourier volumes are arbitrary
+        channel x height x width complex valued tensors,
+        they are not assumed to be Fourier transforms of a real valued 3D functions.
+
+        Note that the tomographic projection is interpolated on a rotated 2D grid.
+        The rotated 2D grid extends outside the boundaries of the 3D grid.
+        The values outside the boundaries are not defined in a useful way.
+        Therefore, in most applications, it make sense to apply a radial filter
+        to the sample.
+
+        """
+        rotmat = rot_params["rotmat"]
+        batch_sz = rotmat.shape[0]
+
+        rotmat = torch.transpose(rotmat, -1, -2)
+        rot_vol_coords = self.vol_coords.repeat((batch_sz, 1, 1)).bmm(rotmat[:, :2, :])
+
+        # rescale the coordinates to be compatible with the edge alignment of
+        # torch.nn.functional.grid_sample
+        if self.config.side_len % 2 == 0:  # even case
```

Comment:
The comments `even case` and ` odd case` might not be necessary as it is evident from the code.